src/source/Response.c

```diff
@@ -93,12 +97,29 @@ STATUS freeCurlResponse(PCurlResponse* ppCurlResponse)
         pCallbacksProvider = pCurlResponse->pCurlRequest->pCurlApiCallbacks->pCallbacksProvider;
     }
 
-    if (pCallbacksProvider != NULL && IS_VALID_MUTEX_VALUE(pCurlResponse->lock)) {
-        locked = pCallbacksProvider->clientCallbacks.tryLockMutexFn(pCallbacksProvider->clientCallbacks.customData, pCurlResponse->lock);
-        // if it fail to acquire the lock that means another thread is tearing down the response already.
-        // thus current thread can return early.
-        // TODO: eliminate the possibility of two threads calling freeCurlResponse.
-        CHK(locked, retStatus);
+    if (pCallbacksProvider != NULL) {
+        if (IS_VALID_MUTEX_VALUE(pCurlResponse->lock)) {
+            locked = pCallbacksProvider->clientCallbacks.tryLockMutexFn(pCallbacksProvider->clientCallbacks.customData, pCurlResponse->lock);
+            // if it fail to acquire the lock that means another thread is tearing down the response already.
+            // thus current thread can return early.
+            // TODO: eliminate the possibility of two threads calling freeCurlResponse.
+            CHK(locked, retStatus);
+        }
+        if (IS_VALID_MUTEX_VALUE(pCurlResponse->dataAvailableLock)) {
+            dataAvailableLocked = pCallbacksProvider->clientCallbacks.tryLockMutexFn(pCallbacksProvider->clientCallbacks.customData, pCurlResponse->dataAvailableLock);
+            CHK(dataAvailableLocked, retStatus);
+
+            if (dataAvailableLocked) {

```

Comment:
If the lock was locked then the entity being signaled is not actively awaiting on the CVAR.

Reply:
Not sure I follow.  https://pubs.opengroup.org/onlinepubs/7908799/xsh/pthread_cond_wait.html

The wait is happening in postReadCallback, under the lock, but when we get to the CVAR_WAIT we will actually release the mutex and block, so yes it is possible for it to be actively waiting on the CVAR once I acquire this lock.

Reply:
What will happen is it will wake up, realize it can't acquire the lock and just block forever or timeout, I see.  That was only there for testing purposes in the middle of some changes right now.